common/decompressingWriter.go

```diff
@@ -0,0 +1,138 @@
+// Copyright © Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"errors"
+	"io"
+	"time"
+)
+
+type decompressingWriter struct {
+	pipeWriter  *io.PipeWriter
+	workerError chan error
+}
+
+const decompressingWriterCopyBufferSize = 256 * 1024 // 1/4 the size that we usually write to disk with (elsewhere in codebase). 1/4 to try to keep mem usage a bit lower, without going so small as to compromize perf
+var decompressingWriterBufferPool = NewMultiSizeSlicePool(decompressingWriterCopyBufferSize)
+
+var ECompressionType = CompressionType(0)
+
+type CompressionType uint8
+
+func (CompressionType) None() CompressionType        { return CompressionType(0) }
+func (CompressionType) ZLib() CompressionType        { return CompressionType(1) }
+func (CompressionType) GZip() CompressionType        { return CompressionType(2) }
+func (CompressionType) Unsupported() CompressionType { return CompressionType(255) }
+
+// NewDecompressingWriter returns a WriteCloser which decompresses the data
+// that is written to it, before passing the decompressed data on to a final destination
+func NewDecompressingWriter(destination io.WriteCloser, ct CompressionType) io.WriteCloser {
```

Comment:
This feels wrong wrapping around another writer in a parasitic way, as it's harder to actually wrap my head around.

I understand completely what it's doing, but for some reason my gut wants it to take in a reader, not a writer, since this is effectively just a pipe through a decompressor.

Reply:
That being said though, this isn't me saying "don't do this"-- I just want to put my thoughts on this out there. Feel free to ping me on teams if you'd like to chat about it.

---
In reply to: [322487718](https://github.com/Azure/azure-storage-azcopy/pull/582#discussion_r322487718) [](ancestors = 322487718)

Reply:
Yeah, I know the feeling.  It too me a while to get things straight in my head. Where I ended up was that it has to be an instance of the decorator pattern - i.e. have the same API as the thing that it wraps. That's so that the rest of our code doesn't care whether its dealing with a plain disk file, or this new magic thing that automatically handles the decompression.  The requirement that the rest of the code shouldn't know (or care) leads to the requirement that both the wrapper and the "wrappee" must have the same interface - which must be writeCloser because the "wrappee" is a disk file and that's a writeCloser.

So it's not really solving the problem on "what would I write if I just wanted to decompress some stuff", its solving the slightly different problem of "what would I write if I wanted to decompress some stuff _transparently_, with an interface that looks just like an ordinary writable file".

I agree that there are some intuitive hurdles here, but on reflection the decorator approach still feels like the right option for the codebase as a whole.  So I'll mark this as resolved.  Definitely agree that we noticed similar mental hurdles here :-)